src/interrupt/mod.rs

```diff
@@ -0,0 +1,193 @@
+//! activities.
+//!
+//! Hardware interrupts are used by devices to communicate that they require attention from the
+//! operating system, or a bare-metal program running on the CPU if there are no OSes. The act of
+//! initiating a hardware interrupt is referred to as an interrupt request (IRQ). Different devices
+//! are usually associated with different interrupts using a unique value associated with each
+//! interrupt. This makes it possible to know which hardware device caused which interrupts.
+//! These interrupt values are often called IRQ lines, or just interrupt lines.
+//!
+//! Nowadays, IRQ lines is not the only mechanism to delivery device interrupts to processors.
+//! MSI [(Message Signalled Interrupt)](https://en.wikipedia.org/wiki/Message_Signaled_Interrupts)
+//! is another commonly used alternative in-band method of signaling an interrupt, using special
+//! in-band messages to replace traditional out-of-band assertion of dedicated interrupt lines.
+//! While more complex to implement in a device, message signaled interrupts have some significant
+//! advantages over pin-based out-of-band interrupt signaling. Message signaled interrupts are
+//! supported in PCI bus since its version 2.2, and in later available PCI Express bus. Some non-PCI
+//! architectures also use message signaled interrupts.
+//!
+//! On the other hand, IRQ is a term commonly used by Operating Systems when dealing with hardware
+//! interrupts. The IRQ number managed by OSes is independent of the IRQ number managed by VMM.
+//! To make things easy, `Interrupt Source` is used instead of IRQ for pin-based interrupts and
+//! MSI interrupts.
+//!
+//! A device may support multiple types of interrupts, and each type of interrupt may support one
+//! or multiple interrupt sources. For example, a PCI device may support:
+//! * Legacy Irq: exactly one interrupt source.
+//! * PCI MSI Irq: 1,2,4,8,16,32 interrupt sources.
+//! * PCI MSIx Irq: 2^n(n=0-11) interrupt sources.
+//!
+//! A distinct Interrupt Source Identifier (ISID) will be assigned to each interrupt source.
+//! An ID allocator will be used to allocate and free Interrupt Source Identifiers for devices.
+//! To decouple the vm-device crate from the ID allocator, the vm-device crate doesn't take the
+//! responsibility to allocate/free Interrupt Source IDs but only makes use of assigned IDs.
+//!
+//! The overall flow to deal with interrupts is:
+//! * the VMM creates an interrupt manager
+//! * the VMM creates a device manager, passing on an reference to the interrupt manager
+//! * the device manager passes on an reference to the interrupt manager to all registered devices
+//! * guest kernel loads drivers for virtual devices
+//! * guest device driver determines the type and number of interrupts needed, and update the
+//!   device configuration
+//! * the virtual device backend requests the interrupt manager to create an interrupt group
```

Comment:
I'm trying to understand how you think this would interact with the [proposed device manager design](https://github.com/rust-vmm/vm-allocator/pull/1#issuecomment-518003480) and update it accordingly:

1. When registering a device, would interrupts be represented as resources at all?

2. With legacy IRQs, the VMM selects the interrupt numbers the device is going to use. Is GSI allocation part of the vm-allocator crate?

3. With MSI, the addresses that the guest chooses are part of the MMIO address space. Do we let the vm-allocator crate know about it, or do we assume there won't be any conflict with other devices where the VMM needs to reserve allocate chunks of the MMIO address space?

Looking forward to getting your input, but my personal answers:

1. Eventually, we will need to have the device manager know about interrupts for any device. In particular for legacy interrupts. Maybe we should only consider legacy interrupts as a device manager resource.

2. I think GSI should be managed by the vm-allocator.

3. That may be part of the next step for updating a device's resources, but here as well the device manager will need to be aware of which MSI addresses have been selected by the guest, and not allocate anything there.

Reply:
> That may be part of the next step for updating a device's resources, but here as well the device manager will need to be aware of which MSI addresses have been selected by the guest, and not allocate anything there.

@chao-p rightfully pointed out that MSI addresses are only used by KVM for resolving the target interrupt. They will not be used for actual IO by the guest. So I guess we can scratch point #3 off.

Reply:
Are you talking about the high_address/low_address in PCI MSI configuration registers? That address must fall into the local APIC address range(0xFEE000000), which is reserved by the VMMs.